apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// An <see cref="IConsumer{T}"/> that will allow for the option to retry RPC failures and buffer
+    /// the data while the retries occur.
+    /// </summary>
+    internal class RpcRetryConsumer<T> : IConsumer<T>
+    {
+        /// <summary>The options dictating how retries will be handled.</summary>
+        private readonly RetryOptions _options;
+
+        /// <summary>A function to obtain the size of an item in bytes.</summary>
+        private readonly Func<T, int> _sizer;
+
+        /// <summary>
+        /// Function to create a <see cref="ISequentialThreadingTimer"/> that will start the given
+        /// action based on retry options.
+        /// </summary>
+        private readonly Func<Action, RetryOptions, ISequentialThreadingTimer> _timerFactory;
+
+        /// <summary>The timer running in the background to retry failed requests.</summary>
+        private readonly ISequentialThreadingTimer _timer;
+
+        /// <summary>A semaphore to protect the buffer.</summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
```

Comment:
It may be simpler, but it is also more dangerous. Using an object, you can use a lock statement to lock on it, ensuring the lock is released when exiting the critical section. The semaphore requires manual management of the critical section.

Case in point: in your changes just now in `FlushBuffer`, you moved the first `_semaphore.Wait()` call above a return statement, so if you bail early you will deadlock on the next call to `FlushBuffer`.